foundation/blockchain/state/state_test.go

```diff
@@ -0,0 +1,332 @@
+package state_test
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ardanlabs/blockchain/foundation/blockchain/database"
+	"github.com/ardanlabs/blockchain/foundation/blockchain/genesis"
+	"github.com/ardanlabs/blockchain/foundation/blockchain/peer"
+	"github.com/ardanlabs/blockchain/foundation/blockchain/state"
+	"github.com/ardanlabs/blockchain/foundation/blockchain/storage/memory"
+	"github.com/ardanlabs/blockchain/foundation/events"
+	"github.com/ardanlabs/blockchain/foundation/logger"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// ================== TOOLKIT FOR TESTS =======================================
+
+const (
+	MINER1_ECDSA   = "8dc79feefd3b86e2f9991def0e5ccd9a5128e104682407b308594bc1032ac7f0"
+	MINER2_ECDSA   = "5aed92a29e1014d83c1d8ac755878723d7e44d8dc129610d11b2022d09ad95bd"
+	MINER3_ECDSA   = "ce07a51ad1d72084aed971b24042f320b4673e852b59eb550375b9eb6747d74a"
+	PERSONA1_ECDSA = "9f332e3700d8fc2446eaf6d15034cf96e0c2745e40353deef032a5dbf1dfed93"
+	PERSONA2_ECDSA = "aed31b6b5a341af8f27e66fb0b7633cf20fc27049e3eb7f6f623a4655b719ebb"
+	PERSONA3_ECDSA = "601d7574860c135e9d3c1d52b0ee997404130edc2a1177c78fda92dd6a3dc2f7"
+)
+
+// Simple worker - required for the proper operation with states,
+// but it does nothing.
+// The production one will work with other goroutines - we dont want this
+// in our test cases.
+type noopWorker struct {
+}
+
+func (n *noopWorker) Shutdown()                              {}
+func (n *noopWorker) Sync()                                  {}
+func (n *noopWorker) SignalStartMining()                     {}
+func (n *noopWorker) SignalCancelMining()                    {}
+func (n *noopWorker) SignalShareTx(blockTx database.BlockTx) {}
+
+// signedTxOpts allows us to simplify creating signed TX, when
+// calling createSignedTX.
+type signedTxOpts struct {
+	keyHex string
+	nonce  uint64
+	to     string
+	value  uint64
+	tip    uint64
+	data   []byte
+}
+
+// createSignedTX - will create 1 signed transaction.
+func createSignedTX(opts signedTxOpts) (database.SignedTx, error) {
```

Comment:
All comments need to be proper sentences, sentence structure, grammar, and puncuation.

Reply:
Fixed.